### PR TITLE
[JENKINS-20679] Try to declare the minimum Java version in the manifest

### PR DESCRIPTION
**Pull request by daniel-beck**
[JENKINS-20679](https://issues.jenkins-ci.org/browse/JENKINS-20679)

Only tested manually, including with a patched plugins POM to see whether the compiler plugin configuration fallback works. Can rip that out and we only support property or HPI plugin configuration for this.

Downstream PRs are https://github.com/jenkins-infra/backend-update-center2/pull/164 and https://github.com/jenkinsci/jenkins/pull/3016.

Upstream of: https://github.com/jenkinsci/plugin-pom/pull/134

@jenkinsci/code-reviewers

### CODE REVIEW

**Comment by KostyaSha** [comment]
Unfortunately hpi-plugin is not using jar-plugin. Does maven-jar-plugin put such info? Could you compare with generic maven project with jar packaging?

**Comment by daniel-beck** [comment]
@KostyaSha Not sure what you mean, could you elaborate? Also, note that I don't want to mechanically pick up whatever is used to build the plugin (may be too recent, or there may be reasons to require a newer JRE to actually run), so the `Build-Jdk` entry is unsuitable.

**Comment by KostyaSha** [comment]
Checked myself. Seems maven-jar is not setting this information. I thought it includes more information in manifest :/

**Comment by daniel-beck** [comment]
@oleg-nenashev Animal sniffer and enforcer use the `java.level` in the plugins POM. While there may be rare situations this isn't good enough (e.g. custom parent POM/overrides), we should just treat those as bugs in the plugin (responsibility to set the HPI plugin parameter if somehow needing a newer JRE as automatically determined) and ignore it otherwise.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java [hunk 1]
```diff
@@ -980,6 +993,42 @@ protected void setAttributes(Section mainSection) throws MojoExecutionException,
...
+     * @return A Java version such as <code>1.7</code> or <code>1.8</code>, or <code>null</code> if it could not be determined.
+     */
+    private String determineMinimumJavaVersion() {
+        // prefer explicitly defined minimum Java version
+        String javaVersion = this.javaVersion;
+
+        // next stop, 2.x plugins parent POM property
+        if (javaVersion == null) {
+            // plugin POM convention
+            javaVersion = this.project.getProperties().getProperty("java.level");
```
Comment (jglick) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java]: No, do not do it this way.

Define a mojo property with no default. Then patch `plugin-pom` to configure it with `${java.level}`.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java [hunk 2]
```diff
@@ -980,6 +993,42 @@ protected void setAttributes(Section mainSection) throws MojoExecutionException,
...
+            if (javaVersion != null && javaVersion.matches("\\d+")) {
+                javaVersion = "1." + javaVersion;
+            }
+        }
+
+        // if all else fails, one last attempt to get it from the maven-compiler-plugin configuration
```
Comment (jglick) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java]: Do not bother. Plugins should use a current POM if they want this feature.

**Comment by daniel-beck** [comment]
This change + the downstream change to plugin POM should probably done sooner rather than later.

The corresponding core change is only needed when core supports Java 9 at all, and the infra change can be implemented even later (theoretically).

**Comment by jglick** [comment]
> the downstream change to plugin POM

Where is that?

File: src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java [hunk 3]
```diff
@@ -118,6 +118,12 @@
     @Parameter
     private String compatibleSinceVersion;
 
+    /**
+     * Optional - specify the minimum version of Java that this plugin requires.
+     */
+    @Parameter
+    private String javaVersion;
```
Comment (batmat) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java]: should this be called `mininumJavaVersion`, as with JDK9, 10 and upcoming, we might have to introduce some `maxJavaVersion` or so? (Hopefully not, but let's still try and embed that "minimum" version implied meaning into the prop name?)

**Comment by jglick** [comment]
Is there something filed downstream of this?

**Comment by daniel-beck** [comment]
@jglick See original PR comment.

**Comment by jglick** [comment]
I think on hold while discussions in the downstreams continue.

**Comment by oleg-nenashev** [comment]
@jenkinsci/java11-support

**Comment by oleg-nenashev** [comment]
@jglick I would like to use the metadata for another use-case in Update Center generation: https://github.com/jenkins-infra/update-center2/pull/253 . Would it be reasonable for you from the downstream justification PoV?

File: src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java [hunk 4]
```diff
@@ -963,6 +969,10 @@ protected void setAttributes(Section mainSection) throws MojoExecutionException,
         if (compatibleSinceVersion!=null)
             mainSection.addAttributeAndCheck(new Attribute("Compatible-Since-Version", compatibleSinceVersion));
 
+        if (this.minimumJavaVersion != null) {
+            mainSection.addAttributeAndCheck(new Attribute("Minimum-Java-Version", this.minimumJavaVersion));
+        }
+
```
Comment (batmat) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java]: Inclined to think we should actually probably issue a warning if not specified, also to advertise that feature.
I strongly fear that we'll have plugins de facto requiring Java 11 to run that will miss they can express this requirement without being nudged about it like here.

Comment (oleg-nenashev) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java]: @batmat I am ready to add validation if it helps.
Speaking of the borderline chekcs in general, https://github.com/jenkinsci/plugin-pom/pull/133 should add some enforcement on the Plugin POM side

**Comment by oleg-nenashev** [comment]
And then Maven integration tests failed... as designed, I believe. Will add flags there

**Comment by oleg-nenashev** [comment]
There is a chicken-and-egg problem in integration tests. They fail, because `minimumJavaVersion` is now required as the reviewers suggested. The problem is that the tests use old versions of Plugin POM which do not include https://github.com/jenkinsci/plugin-pom/pull/134 (downstream of this PR).

*option 1*: I could release a snapshot of the Parent POM, update integration tests to such timestamped snapshot (upper bounds, yey), and then the tests would be green. But it requires . 

*option 2*: Reconsider the approach and do not require `minimumJavaVersion` to be always set. In such case we will have no breaking changes, and the integration tests will pass with previous plugin POMs

@jglick @daniel-beck @jenkinsci/java11-support Would appreciate an advice how to proceed

**Comment by batmat** [comment]
IMO if we step back: 
* in the general case, i.e. if we don't take the current state in account, I am still convinced we should require it. It's much clearer and we can leverage this in many locations to serve better the Jenkins users by filtering out the incompatible plugins for their setup. AFAIK we all agree that this is more and more needed, if not overdue.
* *anyway* nobody is going to see her/his plugin start failing to build right after we merge and release the new `maven-hpi-plugin`. This requires an explicit version bump, and in such case said developers will just spend the needed 5 seconds to decide what minimum Java version they are actually supporting, done.
* The ITs here are obviously a special case, and are failing exactly for the contrary reason as the previous point: this uses the latest version of the `maven-hpi-plugin` on an old plugin. Which in real life would be no-brainer: developer adds the newly required `minimumJavaVersion` field, done. So despite this is a bit crappy here to have to modify existing ITs, I don't see really an other way out, so let's bite the bullet.

So, for the issue at stake here IMO this is why we should go for *option 1*, because requiring the field is the most important and the right long-term decision.

**Comment by jglick** [comment]
Alternately, make the Java version optional (with a warning) for now, release everything, then immediately afterward go back and make it mandatory by switching the parent POMs in the ITs to use the prior release.

**Comment by oleg-nenashev** [comment]
Deployed the current state as `3.0-20181203.145945-2`

**Comment by oleg-nenashev** [comment]
Going forward with option 1

File: src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java (hunk 1 above)
Comment (oleg-nenashev) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java]: done

**Comment by oleg-nenashev** [comment]
@jenkinsci/java11-support @jglick  The rest of integration tests are failing with...

```
[WARNING] Could not validate integrity of download from https://repo.jenkins-ci.org/public/org/jenkins-ci/tools/maven-hpi-plugin/3.0-SNAPSHOT/maven-hpi-plugin-3.0-20181203.145945-2.pom.sha1: Checksum validation failed, no checksums available

[WARNING] Could not validate integrity of download from https://repo.jenkins-ci.org/public/org/jenkins-ci/tools/maven-hpi-plugin/3.0-SNAPSHOT/maven-hpi-plugin-3.0-20181203.145945-2.jar.sha1: Checksum validation failed, no checksums available

[INFO] ..FAILED (44.9 s)

```

I cannot do much about that, so I suggest merging as is and then fixing the PR builder after releasing the downstream patch and bumping to the released version of Plugin POM. Are you fine with it?

**Comment by batmat** [comment]
You mean releasing the `plugin-pom` https://github.com/jenkinsci/plugin-pom/pull/134. I guess let's do that, yes.

File: README.md [hunk 5]
```diff
@@ -5,6 +5,15 @@ See the [developer guide](https://jenkins.io/doc/developer/plugin-development/)
 
 ## Changelog
 
+### 3.0 (2018-11-30)
+
+* [JENKINS-20679](https://issues.jenkins-ci.org/browse/JENKINS-20679) - 
+Inject `Minimum-Java-Version` into the manifest.
+  * It is set by a new mandatory `minimumJavaVersion` parameter in `hpi:hpi`, `hpi:jar` and `hpi:hpl`
+  * Format: `java.specification.version` according to [Java JEP-223](https://openjdk.java.net/jeps/223). 
+    Examples: `1.6`, `1.7`, `1.8`, `6`, `7`, `8`, `9`, `11`, ...
```
Comment (jglick) [inline README.md]: Huh? That JEP seems to say the format should be `6`, `7`, etc., and _not_ `1.6`, `1.7`, etc. as before. So which is the version format expected by this feature?

Comment (oleg-nenashev) [reply README.md]: https://openjdk.java.net/jeps/223 . `1.6` and `1.7` are the old formats of Java specification, we should not maintain them going forward imho

**Comment by oleg-nenashev** [comment]
Maven release failed due to the duplicate metadata file, but I see it for other Maven Plugins in Jenkins org as well. The release itself should be fine

```
[INFO] [INFO] --- maven-deploy-plugin:2.6:deploy (default-deploy) @ maven-hpi-plugin ---
[INFO] Uploading to maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-hpi-plugin/3.0/maven-hpi-plugin-3.0.jar
[INFO] Progress (1): 2.0/140 kB
[INFO] Progress (1): 4.1/140 kB
[INFO] Progress (1): 6.1/140 kB
[INFO] Progress (1): 8.2/140 kB
[INFO] Progress (1): 10/140 kB 
[INFO] Progress (1): 12/140 kB
[INFO] Progress (1): 14/140 kB
[INFO] Progress (1): 16/140 kB
[INFO] Progress (1): 18/140 kB
[INFO] Progress (1): 20/140 kB
[INFO] Progress (1): 23/140 kB
[INFO] Progress (1): 25/140 kB
[INFO] Progress (1): 27/140 kB
[INFO] Progress (1): 29/140 kB
[INFO] Progress (1): 31/140 kB
[INFO] Progress (1): 33/140 kB
[INFO] Progress (1): 35/140 kB
[INFO] Progress (1): 37/140 kB
[INFO] Progress (1): 39/140 kB
[INFO] Progress (1): 41/140 kB
[INFO] Progress (1): 43/140 kB
[INFO] Progress (1): 45/140 kB
[INFO] Progress (1): 47/140 kB
[INFO] Progress (1): 49/140 kB
[INFO] Progress (1): 51/140 kB
[INFO] Progress (1): 53/140 kB
[INFO] Progress (1): 55/140 kB
[INFO] Progress (1): 57/140 kB
[INFO] Progress (1): 59/140 kB
[INFO] Progress (1): 61/140 kB
[INFO] Progress (1): 63/140 kB
[INFO] Progress (1): 66/140 kB
[INFO] Progress (1): 68/140 kB
[INFO] Progress (1): 70/140 kB
[INFO] Progress (1): 72/140 kB
[INFO] Progress (1): 74/140 kB
[INFO] Progress (1): 76/140 kB
[INFO] Progress (1): 78/140 kB
[INFO] Progress (1): 80/140 kB
[INFO] Progress (1): 82/140 kB
[INFO] Progress (1): 84/140 kB
[INFO] Progress (1): 86/140 kB
[INFO] Progress (1): 88/140 kB
[INFO] Progress (1): 90/140 kB
[INFO] Progress (1): 92/140 kB
[INFO] Progress (1): 94/140 kB
[INFO] Progress (1): 96/140 kB
[INFO] Progress (1): 98/140 kB
[INFO] Progress (1): 100/140 kB
[INFO] Progress (1): 102/140 kB
[INFO] Progress (1): 104/140 kB
[INFO] Progress (1): 106/140 kB
[INFO] Progress (1): 109/140 kB
[INFO] Progress (1): 111/140 kB
[INFO] Progress (1): 113/140 kB
[INFO] Progress (1): 115/140 kB
[INFO] Progress (1): 117/140 kB
[INFO] Progress (1): 119/140 kB
[INFO] Progress (1): 121/140 kB
[INFO] Progress (1): 123/140 kB
[INFO] Progress (1): 125/140 kB
[INFO] Progress (1): 127/140 kB
[INFO] Progress (1): 129/140 kB
[INFO] Progress (1): 131/140 kB
[INFO] Progress (1): 133/140 kB
[INFO] Progress (1): 135/140 kB
[INFO] Progress (1): 137/140 kB
[INFO] Progress (1): 139/140 kB
[INFO] Progress (1): 140 kB    
[INFO]                     
[INFO] Uploaded to maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-hpi-plugin/3.0/maven-hpi-plugin-3.0.jar (140 kB at 69 kB/s)
[INFO] Uploading to maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-hpi-plugin/3.0/maven-hpi-plugin-3.0.pom
[INFO] Progress (1): 2.0/12 kB
[INFO] Progress (1): 4.1/12 kB
[INFO] Progress (1): 6.1/12 kB
[INFO] Progress (1): 8.2/12 kB
[INFO] Progress (1): 10/12 kB 
[INFO] Progress (1): 12/12 kB
[INFO] Progress (1): 12 kB   
[INFO]                    
[INFO] Uploaded to maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-hpi-plugin/3.0/maven-hpi-plugin-3.0.pom (12 kB at 15 kB/s)
[INFO] Downloading from maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-hpi-plugin/maven-metadata.xml
[INFO] Progress (1): 0.7/2.6 kB
[INFO] Progress (1): 2.1/2.6 kB
[INFO] Progress (1): 2.6 kB    
[INFO]                     
[INFO] Downloaded from maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-hpi-plugin/maven-metadata.xml (2.6 kB at 7.3 kB/s)
[INFO] Downloading from maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-metadata.xml
[INFO] Progress (1): 385 B
[INFO]                    
[INFO] Downloaded from maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-metadata.xml (385 B at 1.5 kB/s)
[INFO] Uploading to maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-hpi-plugin/maven-metadata.xml
[INFO] Progress (1): 2.0/2.5 kB
[INFO] Progress (1): 2.5 kB    
[INFO]                     
[INFO] Uploaded to maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-hpi-plugin/maven-metadata.xml (2.5 kB at 3.1 kB/s)
[INFO] Uploading to maven.jenkins-ci.org: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-metadata.xml
[INFO] Progress (1): 385 B
[INFO]                    
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] BUILD FAILURE
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] Total time:  26.541 s
[INFO] [INFO] Finished at: 2018-12-05T09:58:38+01:00
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:2.6:deploy (default-deploy) on project maven-hpi-plugin: Failed to deploy metadata: Could not transfer metadata org.jenkins-ci.tools/maven-metadata.xml from/to maven.jenkins-ci.org (https://repo.jenkins-ci.org/releases/): Access denied to: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-metadata.xml -> [Help 1]
[INFO] [ERROR] 
[INFO] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[INFO] [ERROR] Re-run Maven using the -X switch to enable full debug logging.
[INFO] [ERROR] 
[INFO] [ERROR] For more information about the errors and possible solutions, please read the following articles:
[INFO] [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by olamy** [comment]
@oleg-nenashev nope not duplicated! but you cannot write the maven-metadata.xml (error message says ` Access denied to: https://repo.jenkins-ci.org/releases/org/jenkins-ci/tools/maven-metadata.xml ` ). I had the same problem recently... might be an infra issue... (not source files are probably not deploy...)

File: README.md (hunk 5 above)
Comment (jglick) [reply README.md]: Right, so why would we support them at all in this newly added feature?

**Comment by daniel-beck** [comment]
The patterns used for upload permissions from https://github.com/jenkins-infra/repository-permissions-updater/ do not allow for a package level metadata file.

**Comment by olamy** [comment]
this is not an issue until we want to add new maven plugin.

**Pull request merged**